Deny Clippy warnings in CI

kangalio opened the pull request:
With this, Rust and Clippy warnings should actually be caught in CI

kangalio committed:
Deny clippy warnings in CI

kangalio commented:
Ah look what's failing the CI: now-functioning Clippy warnings!

kangalio committed:
Fix Clippy warnings

kangalio committed:
Fix clippy

- #[must_use] on shard_id
- Fixed model prelude that doesn't import from deprecated places

kangalio committed:
Fix e16 clippy

mkrasnitski commented:
Rust 1.64 brought some new lints, particularly one showstopper that I already addressed in #2174. But, as it stands, feel free to fix the rest with additional commits to this PR.

EDIT: Some of the lints were actually part of 1.65, my mistake. Right now, on 1.64, the current branch does not trigger clippy.

mkrasnitski commented on src/lib.rs:
```diff
@@ -71,6 +71,8 @@
     clippy::cast_lossless,
     // Covered by other lints
     clippy::missing_panics_doc, // clippy::unwrap_used
+    // Callbacks with `+ 'static + Send + Sync`, trigger this pretty easily
+    clippy::type_complexity,
```
```suggestion
```
I think you should instead define a type alias for the collector like:`type Filter<T> = Arc<dyn Fn(&Arc<T>) -> bool + 'static + Send + Sync>;`, and then just replace the uses of it. The next branch essentially has this, but souped up with GATs and the like.

mkrasnitski commented on src/model/guild/role.rs:
```diff
@@ -307,7 +307,7 @@ impl FromStrAndCache for Role {
 /// [Discord docs](https://discord.com/developers/docs/topics/permissions#role-object-role-tags-structure).
 #[derive(Clone, Debug, Default, Deserialize, Serialize)]
-#[cfg_attr(test, derive(PartialEq))]
+#[cfg_attr(test, derive(PartialEq, Eq))]
```
```suggestion
#[cfg_attr(test, derive(Eq, PartialEq))]
```
Keep these in alphabetical order.

kangalio replied:
I agree with the goal of having consistency, however, serenity doesn't sort derives in lots of places

![Screenshot_20220927_154207](https://user-images.githubusercontent.com/21220820/192542871-4a7786aa-e726-420d-bd25-cecab983d01d.png)

For consistency, all should be fixed, but in a separate PR. But even then, I think it's futile to manually order derives. The slight aesthetic satisfaction is not worth the laborious editing.

Eventually rustfmt should get support to automatically sort derives

mkrasnitski commented:
@kangalioo Thoughts on the suggested changes?

kangalio commented:
Totally forgot about them. Will do later

kangalio committed:
Merge remote-tracking branch 'origin/current' into deny-clippy-warnings-in-ci

kangalio committed:
Remove clippy::type_complexity allow

Because @mkrasnitski aka polarbits disagreed. And also, apparently... the type_complexity lint just isn't triggered anymore now?

kangalio committed:
Nevermind, type_complexity did trigger

I just didn't have the collector feature enabled locally

Feature spaghetti is a bit annoying

kangalio committed:
Warn instead of deny during development

Programmers will still see those lints, and due to `-D warnings` CI will still catch those lints and fail PRs if they trigger them

mkrasnitski commented on src/lib.rs:
```diff
@@ -71,18 +71,8 @@
-    clippy::type_complexity,
 )]
 #![deny(rustdoc::broken_intra_doc_links, rustdoc::private_intra_doc_links)]
-#![deny(
-    clippy::unwrap_used,
-    clippy::non_ascii_literal,
-    clippy::missing_errors_doc,
-    clippy::let_underscore_must_use,
-    clippy::unused_async,
-    unused
-)]
```
I'm not sure we need to remove these entirely. AFAIK, `unwrap_used`, `non_ascii_literal`, and `let_underscore_must_use` are part of `clippy::restriction`, so better to keep them as warnings. Plus, I'm not sure if `clippy::unused_async` is implied by Rust's `unused`, since I think that's part of the compiler itself (do they interface at all?). We need to make sure we're not accidentally allowing any of these by removing them.

kangalio replied:
You're totally correct, thanks 😅

kangalio committed:
Add back accidentally disabled lints

mkrasnitski commented:
I think you missed `warn(unused)`, unless it's implied by something else.

kangalio committed:
Add #[warn(unused)] back in

I thought it's warn-by-default but... of the 21 lints in the unused group, 19 are warn-by-default, but the remaining two, unused_extern_crates and unused_macro_rules, are allow-by-default for no clear reason. Probably an oversight by Rust. I don't feel like opening an issue though

kangalio committed:
Merge remote-tracking branch 'origin/current' into deny-clippy-warnings-in-ci

mkrasnitski commented on src/lib.rs:
```diff
@@ -72,14 +72,7 @@
-    clippy::unused_async,
-    unused
-)]
+#![warn(clippy::unwrap_used, clippy::non_ascii_literal, clippy::let_underscore_must_use, unused)]
```
One final nit, you may want to combine this with the `#![warn(clippy::pedantic)]` up above.

kangalio committed:
Merge #![warn]'s

And move them to the front, else the blanket warnings would override the individual allow's

mkrasnitski commented on src/lib.rs:
```diff
@@ -135,6 +140,7 @@ impl AsRef<Cache> for CacheAndHttp {
 #[cfg(feature = "client")]
 impl AsRef<Http> for CacheAndHttp {
     fn as_ref(&self) -> &Http {
+        let _ = 12_u64 as u8;
```
Missed a spot ;)

kangalio replied:
Oh crap I wanted to verify my "else the blanket warnings would override the individual allow's" claim but forgot to take the test code out :sweat_smile: Thanks for your rigorous reviewing

kangalio committed:
Remove accidentally-commited test code

mkrasnitski commented on examples/e16_sqlite_database/src/main.rs:
```diff
@@ -1,6 +1,8 @@
 // It is recommended that you read the README file, it is very important to this example.
 // This example will help us to use a sqlite database with our bot.
 
+use std::fmt::Write as _;
```
```suggestion
use std::fmt::Write;
```

Ok, last one for real

kangalio replied:
Why? The Write trait is not referenced anywhere, we only need it in scope for `writeln!` to work

mkrasnitski replied:
I think these have the same effect, right? Technically, the first one means you can't name the trait explicitly in code, but I don't think it's a problem, plus it looks a bit messy.

kangalio replied:
I think this is entirely subjective

To make sure, I asked on the community server (https://discord.com/channels/273534239310479360/818964227783262209/1025179266209562654) and they could find no strong objective reason for one or the other. The tendency was towards `as _` to avoid conflicts if the user copy-pastes the example, adds file IO and imports `std::io::Write`.

If there are no other objections I'd keep this line as is?

mkrasnitski replied:
Ok, makes sense. I won't belabor the point further.

mkrasnitski commented:
LGTM!
